src/main/java/com/android/volley/RequestQueue.java

```diff
@@ -38,11 +41,56 @@
     // TODO: This should not be a generic class, because the request type can't be determined at
     // compile time, so all calls to onRequestFinished are unsafe. However, changing this would be
     // an API-breaking change. See also: https://github.com/google/volley/pull/109
+    @Deprecated // Use RequestEventListener instead.
     public interface RequestFinishedListener<T> {
         /** Called when a request has finished processing. */
         void onRequestFinished(Request<T> request);
     }
 
+    /** Request event types listeners will be notified about. */
+    @Retention(RetentionPolicy.SOURCE)
+    @IntDef({
+        RequestEvent.REQUEST_QUEUED,
+        RequestEvent.REQUEST_CACHE_LOOKUP_STARTED,
+        RequestEvent.REQUEST_CACHE_LOOKUP_FINISHED,
+        RequestEvent.REQUEST_PROCESSING_STARTED,
+        RequestEvent.REQUEST_PROCESSING_FINISHED,
+        RequestEvent.REQUEST_FINISHED
+    })
+    public @interface RequestEvent {
+        /** The request was added to the queue. */
+        public static final int REQUEST_QUEUED = 0;
+        /** The cache lookup started for the request. */
+        public static final int REQUEST_CACHE_LOOKUP_STARTED = 1;
+        /**
+         * The cache lookup finished for the request and cached response is delivered or request is
```

Comment:
nit: "and the cached response is delivered or the request is queued for network dispatching."

Reply:
I haven't had a free minute to review this today and I'll be at an offsite for most of tomorrow. I'll try to get to it in the morning but it may be Monday before I can give it a full review. I hope that's okay.